### PR TITLE
CRUD Subject

### PR DESCRIPTION
**Pull request by vjnhat**

### CODE REVIEW

File: framgia/models/category.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__author__ = 'nhat'
```
Comment (phuclc) [inline framgia/models/category.py]: Delete file category.py
file

File: framgia/models/course.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
 __author__ = 'FRAMGIA\le.cong.phuc'
 from django.db import models
-from . import subject
+from . import subject, category
```
Comment (phuclc) [inline framgia/models/course.py]: Delete import category

File: framgia/models/course.py [hunk 3]
```diff
@@ -17,13 +17,14 @@ class Course(models.Model):
...
         return  self.name
 
-    def get_subjects_list(self):
-        return self.subjects.all()
 
     def get_absolute_url(self):
         from django.core.urlresolvers import reverse
         return reverse('CourseDetail.as_view()' , args=[str(self.id)])
 
+    # def get_all_subject(self):
```
Comment (phuclc) [inline framgia/models/course.py]: Pls remove code if you don't use

File: framgia/models/subject.py [hunk 4]
```diff
@@ -12,8 +11,23 @@ class Subject(models.Model):
     end_at = models.DateTimeField('End at', default= timezone.now()+ datetime.timedelta(days = 90))
     status = models.BooleanField('Status', default= True)
 
+
```
Comment (phuclc) [inline framgia/models/subject.py]: Pep8: One blank line before a function

File: framgia/models/subject.py [hunk 5]
```diff
@@ -12,8 +11,23 @@ class Subject(models.Model):
...
+        list_course = []
+        for course in Course.objects.all():
+            if course in list_other:
+               continue
+            else:
+                list_course.append(course)
+        return list_course
+
+
+
```
Comment (phuclc) [inline framgia/models/subject.py]: Pep8: Two blank line before a class

File: framgia/models/subject.py [hunk 6]
```diff
@@ -12,8 +11,23 @@ class Subject(models.Model):
...
     def __str__(self):
         return self.name
 
+    """ get courses """
+    def get_course(self):
+        from framgia.models.course import Course
+        list_other = Course.objects.exclude(subjects = self)
+        list_course = []
+        for course in Course.objects.all():
+            if course in list_other:
```
Comment (phuclc) [inline framgia/models/subject.py]: You should use if course not in list_other: list_course.append(course). Don't use continue command

File: framgia/models/course.py [hunk 7]
```diff
@@ -31,5 +32,23 @@ class Meta:
...
+    def delete(self, course, subject):
+        """
+
+        :param course: class: Course, var: course
+        :param subject: class : Subject, var: subject
+        :return: delete record has subject = subject, course = course
+        it mean delete relationship between a course and a  subject
+        """
     class Meta:
         app_label = 'framgia'
```
Comment (phuclc) [inline framgia/models/course.py]: Need new line at end of file

File: framgia/models/subject.py [hunk 8]
```diff
@@ -12,8 +11,23 @@ class Subject(models.Model):
...
+               continue
+            else:
+                list_course.append(course)
+        return list_course
+
+
+
     class Meta:
         app_label = 'framgia'
```
Comment (phuclc) [inline framgia/models/subject.py]: Need new line at end of file

File: framgia/templates/admin/course/detail.html [hunk 9]
```diff
@@ -4,4 +4,11 @@
...
         <li>Description: {{ course.description }}</li>
     </ul>
+    <ul>Subjects in {{ course.name }}
+        {% for subject in course.subjects.all %}
+            <a href="{% url 'subject_detail' subject.id %}">
+                <li> {{ subject.name }}</li>
+            </a>
+        {% endfor %}
+    </ul>
 {% endblock %}
```
Comment (phuclc) [inline framgia/templates/admin/course/detail.html]: Need new line at end of file

File: framgia/templates/admin/course/form/course_form.html [hunk 10]
```diff
@@ -1,7 +1,14 @@
...
     <form action="" method="post">{% csrf_token %}
-        {{ form.as_p }}
+        <ul>
+            <li> {{ form.as_p }}</li>
+        {% endfor %}
+        </ul>
+
         <input type="submit" value="Create" />
     </form>
 {% endblock %}
```
Comment (phuclc) [inline framgia/templates/admin/course/form/course_form.html]: Need new line at end of file

File: framgia/templates/admin/subject/detail.html [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    </ul>
+    <ul>{{ subject.name }} belongs to courses:
+        {% for course in subject.get_course %}
+            <li>
+                <a href="{% url 'course_detail' course.id %}"> {{ course.name }}</a>
+            </li>
+        {% endfor %}
+    </ul>
+    
+{% endblock %}
```
Comment (phuclc) [inline framgia/templates/admin/subject/detail.html]: Need new line at end of file

File: framgia/templates/admin/subject/index.html [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+                <a href= "{%  url 'subject_delete' object.id %}"><i> Delete </i></a>
+            </li>
+
+            <li>Description:{{ object.description }}</li>
+        </ul>
+    {% endfor %}
+     </center>
+    <a href="{% url 'subject_new' %}"> <b> New Subject</b> </a>
+
+ {% endblock %}
```
Comment (phuclc) [inline framgia/templates/admin/subject/index.html]: Need new line at end of file

File: framgia/templates/admin/subject/subject_form.html [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{% extends 'base.html' %}
+{% block content %}
+    <form action="" method="post">{% csrf_token %}
+        <ul>
+            <li> {{ form.as_p }}</li>
+        </ul>
+        <input type="submit" value="Create" />
+    </form>
+{% endblock %}
```
Comment (phuclc) [inline framgia/templates/admin/subject/subject_form.html]: Need new line at end of file

File: framgia/templates/admin/subject/subject_update_form.html [hunk 14]
```diff
@@ -0,0 +1,9 @@
+{% extends 'base.html' %}
+{% block content %}
+    <form action="" method="post">{% csrf_token %}
+        <ul>
+            <li> {{ form.as_p }}</li>
+        </ul>
+        <input type="submit" value="Create" />
+    </form>
+{% endblock %}
```
Comment (phuclc) [inline framgia/templates/admin/subject/subject_update_form.html]: Need new line at end of file

File: framgia/views/admin/subject.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+class SubjectDelete(DeleteView):
+    """ Delete a subject """
+    model = Subject
+    success_url = reverse_lazy('subject_index')
+    template_name = 'admin/subject/subject_confirm_delete.html'
+
+
+class SubjectCreate(CreateView):
+    """ Create a new subject
```
Comment (phuclc) [inline framgia/views/admin/subject.py]: If a comment is not too long, it should be on one line

File: framgia/views/admin/subject.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    template_name = "admin/subject/index.html"
+
+
+class SubjectDetail(DetailView):
+    """ display all information about a subject such as name, description, courses."""
+    model = Subject
+    template_name = 'admin/subject/detail.html'
+
+
+
```
Comment (phuclc) [inline framgia/views/admin/subject.py]: Pep8: Two blank line before a class

File: framgia/views/admin/course.py [hunk 17]
```diff
@@ -38,6 +38,8 @@ class CourseCreate(CreateView):
     """ Create a new course
```
Comment (phuclc) [inline framgia/views/admin/course.py]: If a comment is not too long, it should be on one line

File: framgia/models/subject.py (hunk 4 above)
Comment (khanhkhanh) [inline framgia/models/subject.py]: Remove 13th or 14th line. Need one line break, not two.

File: framgia/models/subject.py (hunk 5 above)
Comment (khanhkhanh) [inline framgia/models/subject.py]: Remove one of (28th - 30th) lines

File: framgia/models/subject.py [hunk 18]
```diff
@@ -12,8 +11,23 @@ class Subject(models.Model):
...
+                list_course.append(course)
+        return list_course
+
+
+
     class Meta:
-        app_label = 'framgia'
\ No newline at end of file
+        app_label = 'framgia'
+
```
Comment (khanhkhanh) [inline framgia/models/subject.py]: Remove tab

File: framgia/templates/admin/course/detail.html [hunk 19]
```diff
@@ -4,4 +4,11 @@
...
-{% endblock %}
\ No newline at end of file
+    <ul>Subjects in {{ course.name }}
+        {% for subject in course.subjects.all %}
+            <a href="{% url 'subject_detail' subject.id %}">
+                <li> {{ subject.name }}</li>
+            </a>
+        {% endfor %}
+    </ul>
+{% endblock %}
```
Comment (khanhkhanh) [inline framgia/templates/admin/course/detail.html]: Use 2-space indent for html, css, js files

File: framgia/templates/admin/course/form/course_form.html [hunk 20]
```diff
@@ -1,7 +1,14 @@
...
+        <ul>
+            <li> {{ form.as_p }}</li>
+        {% endfor %}
+        </ul>
+
         <input type="submit" value="Create" />
     </form>
-{% endblock %}
\ No newline at end of file
+{% endblock %}
```
Comment (khanhkhanh) [inline framgia/templates/admin/course/form/course_form.html]: Use 2-space indent for html, css, js files

File: framgia/templates/admin/course/form/course_form.html [hunk 21]
```diff
@@ -1,7 +1,14 @@
 {% extends 'base.html' %}
+{% block title %}
+    Create a new course
+{% endblock %}
 {% block content %}
     <form action="" method="post">{% csrf_token %}
-        {{ form.as_p }}
+        <ul>
```
Comment (khanhkhanh) [inline framgia/templates/admin/course/form/course_form.html]: {% csrf_token %} should be in a separate line

File: framgia/templates/admin/course/form/course_form.html [hunk 22]
```diff
@@ -1,7 +1,14 @@
 {% extends 'base.html' %}
+{% block title %}
+    Create a new course
```
Comment (khanhkhanh) [inline framgia/templates/admin/course/form/course_form.html]: Try to learn and use i18n in Django

File: framgia/views/admin/subject.py (hunk 16 above)
Comment (khanhkhanh) [inline framgia/views/admin/subject.py]: Remove one of (20th - 22th) lines

File: framgia/views/admin/course.py [hunk 23]
```diff
@@ -35,9 +34,9 @@ class CourseDelete(DeleteView):
 
 
 class CourseCreate(CreateView):
-    """ Create a new course
-    """
+    """ Create a new course """
     model = Course
+    lst_subjects = Subject.objects.all()
```
Comment (khanhkhanh) [inline framgia/views/admin/course.py]: Remove if not use

File: framgia/templates/admin/subject/index.html [hunk 24]
```diff
@@ -0,0 +1,21 @@
+{% extends 'base.html' %}
+{% block title %}
+    Subject Index
+{% endblock %}
+ {% block content %}
+     <center>
```
Comment (khanhkhanh) [inline framgia/templates/admin/subject/index.html]: Indent

File: framgia/templates/admin/subject/index.html [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+    <h2> List Subjects</h2>
```
Comment (khanhkhanh) [inline framgia/templates/admin/subject/index.html]: Indent

File: framgia/templates/admin/course/index.html [hunk 26]
```diff
@@ -19,4 +19,4 @@
      </center>
     <a href="{% url 'course_new' %}"> <b> New Course</b> </a>
 
- {% endblock %}
\ No newline at end of file
+ {% endblock %}
```
Comment (khanhkhanh) [inline framgia/templates/admin/course/index.html]: Indent for this file

File: framgia/templates/admin/subject/index.html [hunk 27]
```diff
@@ -0,0 +1,21 @@
+{% extends 'base.html' %}
+{% block title %}
+    Subject Index
+{% endblock %}
+ {% block content %}
```
Comment (khanhkhanh) [inline framgia/templates/admin/subject/index.html]: Indent

**Pull request merged**